### DISCUSSION TITLE
System.Text.Json Deserializing Nested Recursive Strings Problem

### DISCUSSION THREAD

**Discussion by bouyeijiang** [dotnet/runtime#104144] [Q&A]
### Description

Incorrect deserialization of nested recursive word JSON string

### Reproduction Steps

 var diagramJson=JsonSerializer.Deserialize<DiagramJson>(jsonString);

### Expected behavior

{
	"id": "DG1719391036524",
	"type": "bpmn:Process",
	"children": [
		{
			"id": "Event_01c21t5",
			"type": "bpmn:StartEvent",
			"roletype": "role",
			"roles": "16b0b2cf-5fa9-4dc4-bd74-81dbee10cc44",
			"children": [
				{
					"id": "Activity_1as9nwc",
					"type": "bpmn:Task",
					"roletype": "role",
					"direction": "1",
					"roles": "cd4104d9-c73b-478a-b116-868158636b3c",
					"children": [
						{
							"id": "Activity_04de6if",
							"type": "bpmn:Task",
							"roletype": "role",
							"direction": "1",
							"roles": "2f814993-9ebd-44d5-9d2e-bc0297227df0",
							"children": [
								{
									"id": "Event_06oarny",
									"type": "bpmn:EndEvent",
									"roletype": "role",
									"direction": "1",
									"endtype": "1"
								}
							]
						},
						{
							"id": "Activity_0hhw92j",
							"type": "bpmn:Task",
							"roletype": "persion",
							"direction": "1",
							"users": "07e8f966-2da9-ede3-847a-693d08991c58",
							"dept": "4a2535de-b3ba-40a8-8cd2-9897ca10603b",
							"children": [
								{
									"id": "Event_06oarny",
									"type": "bpmn:EndEvent",
									"roletype": "role",
									"direction": "1",
									"endtype": "1"
								}
							]
						}
					]
				}
			]
		}
	]
}

### Actual behavior

Only two layers of tree depth can be deserialized

**Comment by dotnet-policy-service** [comment]
Tagging subscribers to this area: @dotnet/area-system-text-json, @gregsdennis
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.

**Comment by bouyeijiang** [comment]
谢谢您的来信！我会及时查阅的

**Comment by huoyaoyuan** [comment]
Can you provide more detail of this issue?

What's the exact content of the `jsonString`? You can print it in either `\"` escaped or non-escaped form.

`JsonSerializer.Deserialize` can't be called with a string only. The result type is required to be passed with type parameter or `Type`/`JsonTypeInfo` object. Are you using strongly-typed result instead of `object`? If so, can you share the definition of the type?

**Comment by bouyeijiang** [comment]
I have provided json string. Anyway, directly deserializing the result is not the desired value.
json string example:

{
	"id": "DG1719391036524",
	"type": "bpmn:Process",
	"children": [
		{
			"id": "Event_01c21t5",
			"type": "bpmn:StartEvent",
			"roletype": "role",
			"roles": "16b0b2cf-5fa9-4dc4-bd74-81dbee10cc44",
			"children": [
				{
					"id": "Activity_1as9nwc",
					"type": "bpmn:Task",
					"roletype": "role",
					"direction": "1",
					"roles": "cd4104d9-c73b-478a-b116-868158636b3c",
					"children": [
						{
							"id": "Activity_04de6if",
							"type": "bpmn:Task",
							"roletype": "role",
							"direction": "1",
							"roles": "2f814993-9ebd-44d5-9d2e-bc0297227df0",
							"children": [
								{
									"id": "Event_06oarny",
									"type": "bpmn:EndEvent",
									"roletype": "role",
									"direction": "1",
									"endtype": "1"
								}
							]
						},
						{
							"id": "Activity_0hhw92j",
							"type": "bpmn:Task",
							"roletype": "persion",
							"direction": "1",
							"users": "07e8f966-2da9-ede3-847a-693d08991c58",
							"dept": "4a2535de-b3ba-40a8-8cd2-9897ca10603b",
							"children": [
								{
									"id": "Event_06oarny",
									"type": "bpmn:EndEvent",
									"roletype": "role",
									"direction": "1",
									"endtype": "1"
								}
							]
						}
					]
				}
			]
		}
	]
}

**Comment by huoyaoyuan** [comment]
The json string is not enough. You are not providing a concrete invocation of the method, and what's the different from expectation. Are you calling `JsonSerializer<object>(jsonString)` or `JsonSerializer<MyModel>(jsonString)`, or something else?

**Comment by bouyeijiang** [comment]
Okay, in the following two lines of code, using the Newtonsoft.Json method yields the expected result, while using System.Text. Json is not what I expected.

  var json=Newtonsoft.Json.JsonConvert.DeserializeObject<DiagramJson>(jsonString);
  var njson = System.Text.Json.JsonSerializer.Deserialize<DiagramJson>(jsonString);

**Comment by huoyaoyuan** [comment]
> var njson = System.Text.Json.JsonSerializer.Deserialize(jsonString);

This line does not compile. There's no overload accepting a single string, without result type information.
Can you check which exact overload are you using?

> Json is not what I expected.

**How** is it unexpected? Is it missing some property or child?

**Comment by bouyeijiang** [comment]
There's no problem with the code,I don't know why gitub lost the <Diagam> code.
![image](https://github.com/dotnet/runtime/assets/4083760/038f8d22-2d9d-440f-a3b1-bb9c3148f30b)

**Comment by huoyaoyuan** [comment]
The generic arguments are interpreted as HTML tags. You need to use code section to avoid this.

Can you share the definition of DiagramJson class then?

**Comment by bouyeijiang** [comment]
oh.

 public class DiagramJson : DiagramBase
 {
     public `List<DiagramChildren>` children { get; set; }
 }

 public class DiagramBase
 {
     public string id { get; set; } = string.Empty;
     public string name { get; set; } = string.Empty;
     public string type { get; set; } = string.Empty;
 }
  
 public class  DiagramChildren:DiagramBase
 {
     public string roletype { get; set; } = string.Empty;
     public string direction { get; set; } = string.Empty;
     public string roles { get; set; } = string.Empty;
     public string users { get; set; } = string.Empty;
     public string dept { get; set;} = string.Empty;
     public `List<DiagramChildren>` children;
 }

**Comment by huoyaoyuan** [comment]
`public List<DiagramChildren> children;` in `DiagramChildren` is a field instead of property. System.Text.Json doesn't include field by default.
You can change the `DiagramChildren` type, or set `JsonSerializerOptions.IncludeFields`.

**Reply by bouyeijiang** [reply]
I see,thank you very much
